File: AlgoExpert/1. Easy/Python/6. Nth Fibonacci/Nth Fibonacci [O(n)].py
# O(n) time | O(n) space

def getNthFib(n, memoize={1: 0, 2: 1}):
	if n in memoize:
		return memoize[n]
	else:
		memoize[n] = getNthFib(n - 1, memoize) + getNthFib(n - 2,memoize)
	return memoize[n]

File: AlgoExpert/1. Easy/C++/6. Nth Fibonacci/Nth Fibonacci [O(n)].cpp
#include <iostream>
#include <unordered_map>

using namespace std;

int getNthFib(int n);

int helper(int n, unordered_map<int, int> memoize);

// O(n) time | O(n) space

int getNthFib(int n) {
	unordered_map<int, int> memoize({{1, 0}, {2, 1}});
	return helper(n, memoize);
}

int helper(int n, unordered_map<int, int> memoize) {
	if (memoize.find(n) != memoize.end()) {
		return memoize[n];
	} else {
		memoize[n] = helper(n - 1, memoize) + helper(n - 2, memoize);
		return memoize[n];
	}
}